zephyr/lib/vpages.c

```diff
@@ -0,0 +1,328 @@
+ * bitmap using the zephyr sys_mem_blocks API.
+ *
+ * The blocks APIs track the allocation of physical memory pages and virtual memory pages
+ * separately. The physical memory pages are mapped to the virtual memory pages as needed to
+ * provide a contiguous virtual memory space to the user.
+ */
+
+#define VPAGE_ALLOC_ELEMS	128	/* max number of allocation elements */
+
+/*
+ * Virtual memory allocation element - tracks allocated virtual page id and size
+ */
+struct valloc_elem {
+	uint16_t pages;		/* number of 4kB pages allocated in contiguous block */
+	uint16_t vpage;		/* virtual page number from start of region */
+};
+
+/*
+ * Virtual page table structure
+ *
+ * This structure holds all information about virtual memory pages
+ * including the number of free and total pages, the virtual memory
+ * region, the block allocator for virtual pages and the allocation
+ * elements.
+ */
+struct vpage_context {
+	struct k_mutex lock;
+	uint32_t free_pages;	/* number of free 4kB pages */
+	uint32_t total_pages;	/* total number of 4kB pages */
+
+	/* Virtual memory region information */
+	const struct sys_mm_drv_region *virtual_region;
+	struct sys_mem_blocks vpage_blocks;
+
+	/* allocation elements to track page id to allocation size */
+	uint32_t num_elems;	/* number of allocated elements in use*/
+	struct valloc_elem velems[VPAGE_ALLOC_ELEMS];
+};
+
+/* uncache persistent across all cores */
+static struct vpage_context page_context;
+static sys_bitarray_t bitmap;
+
+/* singleton across all cores */
+static int vpage_init_done;
+
+/**
+ * @brief Allocate and map virtual memory pages
+ *
+ * Allocates memory pages from the virtual page allocator.
+ * Maps physical memory pages to the virtual region as needed.
+ *
+ * @param pages Number of 4kB pages to allocate.
+ * @param ptr Pointer to store the address of allocated pages.
+ * @retval 0 if successful.
+ */
+static int vpages_alloc_and_map(uint32_t pages, void **ptr)
+{
+	void *vaddr;
+	int ret;
+
+	/* check for valid pages and ptr */
+	if (!pages || !ptr)
+		return -EINVAL;
+
+	*ptr = NULL;
+
+	/* quick check for enough free pages */
+	if (page_context.free_pages < pages) {
+		LOG_ERR("error: not enough free pages %d for requested pages %d",
+			page_context.free_pages, pages);
+		return -ENOMEM;
+	}
+
+	/* check for allocation elements */
+	if (page_context.num_elems >= VPAGE_ALLOC_ELEMS) {
+		LOG_ERR("error: max allocation elements reached");
+		return -ENOMEM;
+	}
+
+	/* allocate virtual continuous blocks */
+	ret = sys_mem_blocks_alloc_contiguous(&page_context.vpage_blocks, pages, &vaddr);
+	if (ret < 0) {
+		LOG_ERR("error: failed to allocate %d continuous virtual pages, free %d",
+			pages, page_context.free_pages);
+		return ret;
+	}
+
+	/* map the virtual blocks in virtual region to free physical blocks */
+	ret = sys_mm_drv_map_region_safe(page_context.virtual_region, vaddr,
+			0, pages * CONFIG_MM_DRV_PAGE_SIZE, SYS_MM_MEM_PERM_RW);
```

Comment:
this is allocating physical pages from anywhere in L2 memory?

Reply:
Free physical pages from L2 SRAM

Reply:
> Free physical pages from L2 SRAM

@lgirdwood right, but shouldn't this be allocating from a pre-allocated set of pages for this "heap" or am I misunderstanding the concept?